Do not create unused testCluster

breskeby opened the pull request:
This avoids creating test clusters that are not required during the build.
We use lazy configuration here on testClusters and only instantiate them as required.

Also updated BUILDING.MD giving hints on using test cluster plugin properly

breskeby committed:
Do not create unused testCluster

This avoids creating test clusters that are not required during the build.
We use lazy configuration here on testClusters and only instantiate them as theyre

breskeby committed:
Do not fail on run task (debug)

breskeby committed:
Create more test cluster lazy

breskeby committed:
Make more test cluster lazy

breskeby committed:
Avoid creating unused testcluster

breskeby committed:
Fix PluginBuildPlugin

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

breskeby committed:
Fix disabling geo db download

breskeby committed:
Fix cluster setup in repository-multi-version

breskeby committed:
Polishing

breskeby committed:
Fix issue with irretic groovy ogic

breskeby committed:
Fix bwc tests

breskeby committed:
Fix more bwcTests

breskeby committed:
Fix more bwc tests

breskeby committed:
Fix more bwc tests

breskeby committed:
Fix more bwc tests

breskeby committed:
Fix typo

breskeby committed:
Minor polishing

breskeby committed:
Fix rolling upgrade tests

breskeby committed:
Fix cluster config in sql qa mixedcluster project

breskeby committed:
Fix more bwc tests

breskeby commented on qa/ccs-rolling-upgrade-remote-cluster/build.gradle:
```diff
@@ -20,8 +20,8 @@ dependencies {
 }
 
 for (Version bwcVersion : BuildParams.bwcVersions.wireCompatible) {
-  String baseName = "v${bwcVersion}"
```
This solves a problem with for loops and lazy evaluated strings that causes late resolved GStrings to only contain the last bwcVersion of the for loop

mark-vieira replied:
Ug, I've been bitten by that so many times.

breskeby committed:
Clean up before review

breskeby committed:
Document test cluster usage

breskeby commented:
@elasticmachine run elasticsearch-ci/part-2

mark-vieira commented on x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:
```diff
@@ -53,11 +50,11 @@ tasks.register("writeJavaPolicy") {
 tasks.register("follow-cluster", RestIntegTestTask) {
   dependsOn 'writeJavaPolicy', "leader-cluster"
-    useCluster testClusters."leader-cluster"
+    useCluster leaderCluster.get()
```
Can we provide an overloaded version of `useCluster` on `TestClusterAware` that accepts a `Provider` so we can avoid all these calls to `get()`?

breskeby replied:
could point. I'll give the testclusters api bit more attention

breskeby replied:
I did some polishing on the api. Besides allow passing a task provider here I also added support for Provider in nonSystemProperties. this makes configurations based on clusters more straight forward by allowing using provider mapping: e.g.

```
tasks.register("someRestIntegTest", StandaloneRestIntegTestTask) {
    useCluster baseCluster
    nonInputProperties.systemProperty('tests.rest.cluster', baseCluster.map(c->c.allHttpSocketURI.join(",")))
}
```

mark-vieira replied:
Nice!

mark-vieira reviewed (approved):
LGTM, although I wish using lazy configured domain objects didn't introduce so much awkwardness in our build scripts.

breskeby committed:
Api polising after Review

provide useCluster(Provider) method to TestClusterAware

Ideally we take this a step further and realize those test clusters only on use.
But out of scope of this PR.

breskeby committed:
Allow gradle provider as value for nonSystemProperties

breskeby committed:
Some simplification on test configuration

breskeby committed:
Fix typo in rest test config

breskeby committed:
Fix more typos

breskeby committed:
Fix another typo

breskeby committed:
Fix more typos

breskeby committed:
Merge branch 'master' into create-less-unused-cluster
